File: src/main/java/org/distril/beengine/player/handler/ResourcePackPacketHandler.java
package org.distril.beengine.player.handler;

import com.nukkitx.protocol.bedrock.BedrockServerSession;
import com.nukkitx.protocol.bedrock.handler.BedrockPacketHandler;
import com.nukkitx.protocol.bedrock.packet.*;
import lombok.extern.log4j.Log4j2;
import org.distril.beengine.material.Material;
import org.distril.beengine.network.Network;
import org.distril.beengine.network.data.LoginData;
import org.distril.beengine.player.Player;
import org.distril.beengine.server.Server;

@Log4j2
public class ResourcePackPacketHandler implements BedrockPacketHandler {

	private final Server server;
	private final BedrockServerSession session;
	private final LoginData loginData;

	public ResourcePackPacketHandler(Server server, BedrockServerSession session, LoginData loginData) {
		this.server = server;
		this.session = session;
		this.loginData = loginData;

		this.session.getHardcodedBlockingId().set(Material.SHIELD.getItemRuntimeId());
		this.session.sendPacket(new ResourcePacksInfoPacket()); // todo: add resource packs
	}

	@Override
	public boolean handle(ResourcePackClientResponsePacket packet) {
		switch (packet.getStatus()) {
			case REFUSED -> this.session.disconnect("disconnectionScreen.noReason");

			case SEND_PACKS -> {
				// todo
			}

			case HAVE_ALL_PACKS -> {
				var stackPacket = new ResourcePackStackPacket();
				stackPacket.setGameVersion(Network.CODEC.getMinecraftVersion());
				this.session.sendPacket(stackPacket);
			}

			case COMPLETED -> {
				this.server.getScheduler().prepareTask(() -> {
					var player = new Player(this.server, this.session, this.loginData);

					this.session.addDisconnectHandler(reason -> {
						this.server.removePlayer(player);

						player.disconnect(reason.name());
					});

					player.initialize();
					player.completePlayerInitialization();

					this.session.setPacketHandler(new PlayerPacketHandler(player));
				}).async(true).schedule();
			}
		}

		return true;
	}

	@Override
	public boolean handle(ResourcePackChunkRequestPacket packet) {
		// todo
		return BedrockPacketHandler.super.handle(packet);
	}

	@Override
	public boolean handle(PacketViolationWarningPacket packet) {
		log.debug("Packet violation: {}", packet);
		return true;
	}
}

File: src/main/kotlin/org/distril/beengine/player/handler/ResourcePackPacketHandler.kt
package org.distril.beengine.player.handler

import com.nukkitx.protocol.bedrock.BedrockServerSession
import com.nukkitx.protocol.bedrock.handler.BedrockPacketHandler
import com.nukkitx.protocol.bedrock.packet.PacketViolationWarningPacket
import com.nukkitx.protocol.bedrock.packet.ResourcePackClientResponsePacket
import com.nukkitx.protocol.bedrock.packet.ResourcePackStackPacket
import com.nukkitx.protocol.bedrock.packet.ResourcePacksInfoPacket
import org.distril.beengine.material.Material
import org.distril.beengine.network.Network
import org.distril.beengine.network.data.LoginData
import org.distril.beengine.player.Player
import org.distril.beengine.server.Server
import org.distril.beengine.util.Utils.getLogger

class ResourcePackPacketHandler(
	private val session: BedrockServerSession,
	private val loginData: LoginData
) : BedrockPacketHandler {

	init {
		this.session.hardcodedBlockingId.set(Material.SHIELD.itemRuntimeId)
		this.session.sendPacket(ResourcePacksInfoPacket()) // todo: add resource packs
	}

	override fun handle(packet: ResourcePackClientResponsePacket): Boolean {
		when (packet.status) {
			ResourcePackClientResponsePacket.Status.SEND_PACKS -> TODO()

			ResourcePackClientResponsePacket.Status.HAVE_ALL_PACKS -> {
				val stackPacket = ResourcePackStackPacket()
				stackPacket.gameVersion = Network.CODEC.minecraftVersion

				this.session.sendPacket(stackPacket)
			}

			ResourcePackClientResponsePacket.Status.COMPLETED -> {
				Server.scheduler.scheduleTask(async = true) {
					val player = Player(this.session, this.loginData)
					this.session.apply {
						this.addDisconnectHandler {
							Server.removePlayer(player)
							player.disconnect(it.name)
						}

						this.packetHandler = PlayerPacketHandler(player)
					}

					player.initialize()
				}
			}

			else -> this.session.disconnect("disconnectionScreen.noReason")
		}

		return true
	}

	override fun handle(packet: PacketViolationWarningPacket): Boolean {
		log.debug("Packet violation: {}", packet)
		return true
	}

	companion object {

		private val log = ResourcePackPacketHandler.getLogger()
	}
}
